rust/x509/src/x509_example_rust/src/wasm_only.rs

```diff
@@ -0,0 +1,513 @@
+use ic_cdk::api::management_canister::ecdsa as cdk_ecdsa;
+use ic_cdk::export_candid;
+use ic_cdk::{api::time, init, update};
+use pkcs8::AssociatedOid;
+use serde::{Deserialize, Serialize};
+use signature::Keypair;
+use spki::{AlgorithmIdentifier, DynSignatureAlgorithmIdentifier, EncodePublicKey};
+use std::{
+    cell::OnceCell, cell::RefCell, convert::TryFrom, convert::TryInto, error::Error, ops::Deref,
+    str::FromStr, time::Duration,
+};
+use x509_cert::{
+    builder::{Builder, CertificateBuilder, Profile},
+    name::Name,
+    request::CertReq,
+    serial_number::SerialNumber,
+    spki::SubjectPublicKeyInfoOwned,
+    time::{Time, Validity},
+};
+
+mod signer;
+use signer::{EcdsaSecp256k1Signer, Ed25519Signer, Sign};
+
+type CanisterId = Principal;
+
+impl TryFrom<&CaKeyInformation> for SchnorrKeyId {
+    type Error = String;
+
+    fn try_from(value: &CaKeyInformation) -> Result<Self, Self::Error> {
+        match value {
+            CaKeyInformation::Ed25519(key_name) => Ok(SchnorrKeyId {
+                algorithm: SchnorrAlgorithm::Ed25519,
+                name: String::from(<&'static str>::from(key_name)),
+            }),
+            something_else => Err(format!(
+                "Expected Ed25519 CA key but got {something_else:?}"
+            )),
+        }
+    }
+}
+
+impl TryFrom<&CaKeyInformation> for cdk_ecdsa::EcdsaKeyId {
+    type Error = String;
+
+    fn try_from(value: &CaKeyInformation) -> Result<Self, Self::Error> {
+        match value {
+            CaKeyInformation::EcdsaSecp256k1(key_name) => Ok(cdk_ecdsa::EcdsaKeyId {
+                curve: cdk_ecdsa::EcdsaCurve::Secp256k1,
+                name: String::from(<&'static str>::from(key_name)),
+            }),
+            something_else => Err(format!(
+                "Expected EcdsaSecp256k1 CA key but got {something_else:?}"
+            )),
+        }
+    }
+}
+
+thread_local! {
+    static CA_KEY_INFORMATION: RefCell<CaKeyInformation> = RefCell::new(CaKeyInformation::Ed25519(KeyName::DfxTestKey));
+
+    // cache the public key and certificate to avoid fetching them multiple times
+    static ROOT_CA_PUBLIC_KEY: OnceCell<Vec<u8>> = OnceCell::new();
+    static ROOT_CA_CERTIFICATE_PEM: OnceCell<String> = OnceCell::new();
+
+    static CHILD_CERTIFICATE_SERIAL_NUMBER: RefCell<u32> = RefCell::new(0);
+}
+
+#[init]
+fn init(ca_key_information: CaKeyInformation) {
+    CA_KEY_INFORMATION.with(|value| {
+        *value.borrow_mut() = ca_key_information;
+    });
+}
+
+#[update]
+async fn root_ca_certificate() -> Result<X509CertificateString, String> {
+    // if the certificate is already cached, return it
+
+    if let Some(certificate) = ROOT_CA_CERTIFICATE_PEM.with(|inner| inner.get().map(|v| v.clone()))
+    {
+        return Ok(X509CertificateString {
+            x509_certificate_string: certificate,
+        });
+    }
+
+    // if the certificate is not cached, create it and try to cache it
+
+    let serial_number = SerialNumber::from(0u32);
```

Comment:
RFC 5280 requires the serial number in a cert be greater than zero.

> The serial number MUST be a positive integer assigned by the CA to each certificate.

> Note: Non-conforming CAs may issue certificates with serial numbers that are negative or zero.  Certificate users SHOULD be prepared to gracefully handle such certificates.

Reply:
changed to 1 :)